hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/keys/KeyStorage.java

```diff
@@ -0,0 +1,249 @@
+   * @param keyDirSuffix the suffix to apply to the keys base path
+   * @throws IOException in case when base directory can not be created with the desired permissions, or if the key
+   *                     algorithm is not available for the {@link java.security.KeyFactory}
+   */
+  public KeyStorage(SecurityConfig config, String component, String keyDirSuffix) throws IOException {
+    this(config, component, Paths.get(config.getKeyLocation(component).toString() + keyDirSuffix));
+  }
+
+  private KeyStorage(SecurityConfig config, String component, Path keyPath) throws IOException {
+    if (config.useExternalCACertificate(component)) {
+      privateKeyPath = config.getExternalRootCaPrivateKeyPath();
+      if (!Files.isReadable(privateKeyPath)) {
+        throw new UnsupportedEncodingException("External private key path is not readable: " + privateKeyPath);
+      }
+      publicKeyPath = config.getExternalRootCaPublicKeyPath();
+      if (!Files.isReadable(publicKeyPath)) {
+        throw new UnsupportedEncodingException("External public key path is not readable: " + publicKeyPath);
+      }
+      externalKeysUsed = true;
+    } else {
+      createOrSanitizeDirectory(keyPath);
+      privateKeyPath = keyPath.resolve(config.getPrivateKeyFileName());
+      publicKeyPath = keyPath.resolve(config.getPublicKeyFileName());
+      externalKeysUsed = false;
+    }
+
+    keyCodec = config.keyCodec();
+  }
+
+  /**
+   * Returns the private key stored in the private key file.
+   *
+   * @return PrivateKey the key read from the private key file.
+   * @throws IOException in case the file is unreadable, or decoding the contents from PEM format fails.
+   */
+  public PrivateKey readPrivateKey() throws IOException {
+    LOG.info("Reading private key from {}.", privateKeyPath);
+    try {
+      return keyCodec.decodePrivateKey(readAllBytes(privateKeyPath));
+    } catch (IOException e) {
+      LOG.error("Failed to read the private key.", e);
+      throw e;
+    }
+  }
+
+  /**
+   * Returns a public key from an encoded file.
+   *
+   * @return PublicKey
+   * @throws IOException              - on Error.
+   */
+  public PublicKey readPublicKey() throws IOException {
+    LOG.info("Reading public key from {}.", publicKeyPath);
+    try {
+      return keyCodec.decodePublicKey(readAllBytes(publicKeyPath));
+    } catch (IOException e) {
+      LOG.error("Failed to read the public key.", e);
+      throw e;
+    }
+  }
+
+  public KeyPair readKeyPair() throws IOException {
+    return new KeyPair(readPublicKey(), readPrivateKey());
+  }
+
+  /**
+   * Stores a given private key using the default config options.
+   *
+   * @param key - Key to write to file.
+   * @throws IOException - On I/O failure.
+   */
+  public void storePrivateKey(PrivateKey key) throws IOException {
+    LOG.info("Storing private key to {}.", privateKeyPath);
+    try {
+      storeKey(privateKeyPath, keyCodec.encodePrivateKey(key));
```

Comment:
The current implementation, it will prevent the private and public key from overwrite again if the file is already exists.  I would prefer we keep this logic.